### PR TITLE
support recursive CTEs logical plans in datafusion-proto

### PR DESCRIPTION
**Pull request by leoyvens**
## Rationale for this change

Fixes a gap in support for logical plan serialization to protobuf.

## What changes are included in this PR?

Adds support for serializing `LogicaPlan::RecursiveQuery` to protobuf.

## Are these changes tested?

Yes, there is an unit test.

## Are there any user-facing changes?

This feature is user visible. <strike>Using it effectively is probably tricky due to the need to decode the provider for the CTE work table in the `LogicalExtensionCodec`. But still, this is a start.</strike>

### CODE REVIEW

File: datafusion/proto/tests/cases/roundtrip_logical_plan.rs [hunk 1]
```diff
@@ -2523,3 +2525,75 @@ fn roundtrip_window() {
...
+    let ctx = SessionContext::new();
+    let logical_round_trip =
+        logical_plan_from_bytes_with_extension_codec(&bytes, &ctx, &extension_codec)
+            .unwrap();
+    let output_round_trip = ctx.sql(query).await.unwrap().collect().await.unwrap();
+
+    assert_eq!(format!("{plan:?}"), format!("{logical_round_trip:?}"));
+    assert_eq!(
+        format!("{}", pretty_format_batches(&output).unwrap()),
+        format!("{}", pretty_format_batches(&output_round_trip).unwrap())
```
Comment (jonahgao) [inline datafusion/proto/tests/cases/roundtrip_logical_plan.rs:2597]: This `assert_eq` looks  incorrect. `output_round_trip` and `output` are both the result of `ctx.sql(query)`.
I think `output_round_trip` should be
```rust
let output_round_trip = ctx
        .execute_logical_plan(logical_round_trip)
        .await
        .unwrap()
        .collect()
        .await
        .unwrap()
```

File: datafusion/proto/tests/cases/roundtrip_logical_plan.rs [hunk 2]
```diff
@@ -2523,3 +2525,75 @@ fn roundtrip_window() {
...
+            Ok(Arc::new(table))
+        }
+
+        fn try_encode_table_provider(
+            &self,
+            _table_ref: &TableReference,
+            _node: Arc<dyn TableProvider>,
+            _buf: &mut Vec<u8>,
+        ) -> Result<(), DataFusionError> {
+            Ok(())
```
Comment (jonahgao) [inline datafusion/proto/tests/cases/roundtrip_logical_plan.rs:2569]: It seems that we need to [encode](https://github.com/apache/datafusion/blob/c48f12d27cc60ae5e9c3d0791d553453f0dd7001/datafusion/proto/src/logical_plan/mod.rs#L964) `CteWorkTable` and then recreate a `CteWorkTable` when decoding.🤔

File: datafusion/proto/tests/cases/roundtrip_logical_plan.rs (hunk 1 above)
Comment (leoyvens) [reply datafusion/proto/tests/cases/roundtrip_logical_plan.rs:2551]: You're absolutely right, and the corrected test was in fact failing! Despite the logical plans being equal, outputs were different.

**Comment by leoyvens** [comment]
@jonahgao thank you so much for your review, this PR is in much better shape now.

**Review by jonahgao — APPROVED** [review]
LGTM👍, thanks @leoyvens 

I merged main to get CI passed and use `Arc::clone` to avoid conflict with #13338

**Pull request merged**